plugin/pkg/admission/priority/admission.go

```diff
@@ -0,0 +1,201 @@
+	}
+	if _, isMirrorPod := pod.Annotations[api.MirrorPodAnnotationKey]; isMirrorPod {
+		return nil
+	}
+	// Make sure that the user has not set `priority` at the time of pod creation.
+	if operation == admission.Create && pod.Spec.Priority != nil {
+		return admission.NewForbidden(a, fmt.Errorf("The integer value of priority must not be provided in pod spec. The system populates the value from the given PriorityClass name"))
+	}
+	var priority int32
+	if len(pod.Spec.PriorityClassName) == 0 {
+		dpc, err := p.findDefaultPriorityClass()
+		if err != nil {
+			return fmt.Errorf("Failed to get default priority class: %v", err)
+		}
+		if dpc != nil {
+			priority = dpc.Value
+		} else {
+			priority = scheduling.DefaultPriorityWhenNoDefaultClassExists
+		}
+	} else {
+		// First try to resolve by system priority classes.
+		priority, ok = SystemPriorityClasses[pod.Spec.PriorityClassName]
+		if !ok {
+			// Now that we didn't find any system priority, try resolving by user defined priority classes.
+			pc, err := p.lister.Get(pod.Spec.PriorityClassName)
+			if err != nil {
+				return fmt.Errorf("Failed to get default priority class %s: %v", pod.Spec.PriorityClassName, err)
+			}
+			if pc == nil {
+				return admission.NewForbidden(a, fmt.Errorf("No PriorityClass with name %v was found", pod.Spec.PriorityClassName))
+			}
+			priority = pc.Value
+		}
+	}
+	pod.Spec.Priority = &priority
+	return nil
+}
+
+func (p *priorityPlugin) admitPriorityClass(a admission.Attributes) error {
+	operation := a.GetOperation()
+	pc, ok := a.GetObject().(*scheduling.PriorityClass)
+	if !ok {
+		return errors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
+	}
+	if pc.Value > HighestUserDefinablePriority {
+		return admission.NewForbidden(a, fmt.Errorf("Maximum allowed value of a user defined priority is %v", HighestUserDefinablePriority))
+	}
+	if _, ok := SystemPriorityClasses[pc.Name]; ok {
+		return admission.NewForbidden(a, fmt.Errorf("The name of the priority class is a reserved name for system use only: %v", pc.Name))
+	}
+	// If the new PriorityClass tries to be the default priority, make sure that no other priority class is marked as default.
+	if pc.GlobalDefault {
+		dpc, err := p.findDefaultPriorityClass()
+		if err != nil {
+			return fmt.Errorf("Failed to get default priority class: %v", err)
+		}
+		if dpc != nil {
+			// Throw an error if a second default priority class is being created, or an existing priority class is being marked as default, while another default already exists.
+			if operation == admission.Create || (operation == admission.Update && dpc.GetName() != pc.GetName()) {
+				return admission.NewForbidden(a, fmt.Errorf("PriorityClass %v is already marked as default. Only one default can exist", dpc.GetName()))
+			}
+		}
+	}
+	return nil
+}
+
+func (p *priorityPlugin) findDefaultPriorityClass() (*scheduling.PriorityClass, error) {
```

Comment:
I wonder what the performance impact of this will be, as it may get run a lot. Though actually, maybe the "lister" is operating from a cache already, so this is just doing an iteration in memory, in which case I think it's fine. But probably you should make sure.